File: src/entities/Collection.ts
import {
  Column,
  Entity,
  Index,
  JoinColumn,
  JoinTable,
  ManyToMany,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { User } from './User';
import { CollectionItem } from './CollectionItem';
import { Tags } from './Tags';

@Entity()
export class Collection {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Index()
  @Column({
    nullable: false,
  })
  name: string;

  @Index()
  @Column({
    nullable: true,
  })
  description: string;

  @Index()
  @Column({ nullable: true })
  created: Date;

  @Index()
  @Column()
  is_private: boolean;

  @Index()
  @Column({
    nullable: true,
  })
  banner_url: string;

  @Index()
  @Column('simple-array', { array: true, nullable: true })
  likes: string[];

  @Index()
  @Column({ default: false, select: false })
  deleted: boolean;

  @Index()
  @Column({
    nullable: true,
  })
  transliteration: string;

  @ManyToMany(
    () => CollectionItem,
    (collectionItem) => collectionItem.collections,
  )
  @JoinTable()
  collectionItems: CollectionItem[];

  @Index()
  @Column({
    nullable: false,
    default: 0,
  })
  subscribers_count: number;

  @Index()
  @Column({
    nullable: false,
    default: 0,
  })
  totalPrice: number;

  @Index()
  @Column({
    nullable: false,
    default: 0,
  })
  shippingTotal: number;

  @Index()
  @Column({
    nullable: true,
  })
  share_string: string;

  @ManyToMany(() => Tags, (tag) => tag.collections)
  tags: Tags[];

  @ManyToOne(() => User, (user) => user.login, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'user' })
  user: string;
}

File: internal/models/collection.go
package models

import (
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type CollectionCollectionItemsCollectionItem struct {
	CollectionID     uuid.UUID `gorm:"primaryKey;column:collectionId"`
	CollectionItemID uuid.UUID `gorm:"primaryKey;column:collectionItemId"`
}

type CollectionTagsCollection struct {
	CollectionID uuid.UUID `gorm:"primaryKey;column:collectionId"`
	TagsID       uuid.UUID `gorm:"primaryKey;column:tagsId"`
}

type Collection struct {
	gorm.Model
	Id               uuid.UUID `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	Name             string
	Description      string
	Created          string
	IsPrivate        bool
	BannerUrl        string
	Likes            int
	Deleted          bool
	Transliteration  string
	SubscribersCount int
	TotalPrice       int
	ShippingTotal    int
	ShareString      string
	CollectionItems  []CollectionItem `gorm:"many2many:collection_collection_items_collection_item;constraint:OnDelete:CASCADE;"`
	User             User             `gorm:"constraint:OnDelete:CASCADE;"`
	Tags             []Tags           `gorm:"many2many:collection_tags_collection;constraint:OnDelete:CASCADE;"`
}

func (Collection) TableName() string {
	return "collection"
}

func (CollectionCollectionItemsCollectionItem) TableName() string {
	return "collection_collection_items_collection_item"
}

func (CollectionTagsCollection) TableName() string {
	return "collection_tags_collection"
}
